### PR TITLE
Add `isSynced` status

### PR DESCRIPTION
**Pull request by soanvig**
- refactor: change how aggregate builds events to be more type-safety, disallow empty events creation
- feat: add `isSynced` property to read model

### CODE REVIEW

File: packages/origin-247-certificate/package.json [hunk 1]
```diff
@@ -4,7 +4,7 @@
     "main": "./dist/js/src/index.js",
     "types": "./dist/js/src/index.d.ts",
     "scripts": {
-        "test:unit": "jest --testRegex 'offchain-certificate\\.spec.ts$'",
+        "test:unit": "jest --testRegex '\\.spec.ts$'",
```
Comment (soanvig) [inline packages/origin-247-certificate/package.json:7]: :sunglasses:

File: packages/origin-247-certificate/src/offchain-certificate/certificate.aggregate.ts [hunk 2]
```diff
@@ -3,65 +3,87 @@ import {
...
+        const initialCertificate = this.createCertificate(firstEvent as CertificateIssuedEvent<T>);
+
+        this.certificate = {
+            ...restEvents.reduce(this.apply.bind(this), initialCertificate),
+            isSynced: this.reduceIsSynced(events)
+        };
     }
 
-    public apply(event: ICertificateEvent): void {
+    private apply(certificate: Cert<T>, event: ICertificateEvent): Cert<T> {
```
Comment (soanvig) [inline packages/origin-247-certificate/src/offchain-certificate/certificate.aggregate.ts:45]: ensuring that `apply` returns  something allows TS to properly ensure, that everything is handled in `switch`

File: packages/origin-247-certificate/src/offchain-certificate/certificate.aggregate.ts [hunk 3]
```diff
@@ -3,65 +3,87 @@ import {
...
 import { CertificateErrors } from './errors';
 import { BigNumber } from 'ethers';
 import { compareDates } from '../utils/date-utils';
 import { ICertificateReadModel } from '..';
 
+type Cert<T> = ICertificateReadModel<T>;
+
 export class CertificateAggregate<T> {
-    private certificate: ICertificateReadModel<T> | null = null;
+    private certificate: Cert<T>;
```
Comment (soanvig) [inline packages/origin-247-certificate/src/offchain-certificate/certificate.aggregate.ts:20]: With this new approach we don't have invalid "null" state of aggregate

File: packages/origin-247-certificate/src/offchain-certificate/certificate.aggregate.ts [hunk 4]
```diff
@@ -3,65 +3,87 @@ import {
...
-            .forEach((event) => this.apply(event));
+        const sortedEvents = [...events].sort((a, b) => compareDates(a.createdAt, b.createdAt));
+        const [firstEvent, ...restEvents] = sortedEvents;
+
+        if (!firstEvent) {
+            throw new CertificateErrors.CertificateNoEvents();
+        }
+
+        if (firstEvent.type !== CertificateEventType.Issued) {
+            throw new CertificateErrors.CertificateNotIssuanceEvent(
```
Comment (JonaszJestem) [inline packages/origin-247-certificate/src/offchain-certificate/certificate.aggregate.ts:31]:
```suggestion
            throw new CertificateErrors.CertificateNotIssuedEvent(
```

Comment (JonaszJestem) [reply packages/origin-247-certificate/src/offchain-certificate/certificate.aggregate.ts:31]: or 
```suggestion
            throw new CertificateErrors.FirstCertificateEventIsNotIssuance(
```

**Review by JonaszJestem — APPROVED** [review]

**Pull request merged**